apps/bf_prefix.c

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 1995-2016 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+#include <openssl/bio.h>
+#include "apps.h"
+
+static int prefix_write(BIO *b, const char *out, size_t outl,
+                        size_t *numwritten);
+static int prefix_read(BIO *b, char *buf, size_t size, size_t *numread);
+static int prefix_puts(BIO *b, const char *str);
+static int prefix_gets(BIO *b, char *str, int size);
+static long prefix_ctrl(BIO *b, int cmd, long arg1, void *arg2);
+static int prefix_destroy(BIO *b);
+static long prefix_callback_ctrl(BIO *b, int cmd, BIO_info_cb *fp);
+
+/* #define DEBUG */
+
+static BIO_METHOD *prefix_meth = NULL;
+
+BIO_METHOD *apps_bf_prefix(void)
+{
+    if (prefix_meth == NULL) {
+        if ((prefix_meth =
+             BIO_meth_new(BIO_TYPE_FILTER, "Prefix filter")) == NULL
+            || !BIO_meth_set_destroy(prefix_meth, prefix_destroy)
+            || !BIO_meth_set_write_ex(prefix_meth, prefix_write)
+            || !BIO_meth_set_read_ex(prefix_meth, prefix_read)
+            || !BIO_meth_set_puts(prefix_meth, prefix_puts)
+            || !BIO_meth_set_gets(prefix_meth, prefix_gets)
+            || !BIO_meth_set_ctrl(prefix_meth, prefix_ctrl)
+            || !BIO_meth_set_callback_ctrl(prefix_meth, prefix_callback_ctrl)) {
+            BIO_meth_free(prefix_meth);
+            prefix_meth = NULL;
+        }
+    }
+    return prefix_meth;
+}
+
+static int prefix_destroy(BIO *b)
+{
+    OPENSSL_free(BIO_get_data(b));
+    return 1;
+}
+
+static int prefix_read(BIO *b, char *in, size_t size, size_t *numread)
+{
+    return BIO_read_ex(BIO_next(b), in, size, numread);
+}
+
+static int prefix_write(BIO *b, const char *out, size_t outl,
+                        size_t *numwritten)
+{
+    const char *prefix = (const char *)BIO_get_data(b);
+
+    *numwritten = 0;
+
+    if (BIO_number_written(b) == 0) {
+        size_t dontcare;
+
+        if (!BIO_write_ex(BIO_next(b), prefix, strlen(prefix), &dontcare))
+            return 0;
+    }
+
+    while (*out != '\0') {
+        size_t i;
+        size_t num = 0;
+        char c;
+
+        for (i = 0; (c = out[i]) != '\n' && c != '\0'; i++)
+            ;
+
+        if (c == '\n')
+            i++;
+
+        while (i > 0) {
+
+            if (!BIO_write_ex(BIO_next(b), out, i, &num))
+                return 0;
+
+            out += num;
+            *numwritten += num;
+            i -= num;
```

Comment:
You could make this into a for loop by putting the i variable up at 84.

Reply:
I would rather keep the use of `num` together, if you don't mind...